Program DofMap in RelationshipManager

lindsayad opened the pull request:
We leverage the `_system_type` of `RelationshipManager` to determine
what `DofMap` from what system we should attach

lindsayad committed:
Let RelationshipManager set DofMap using a SubProblem argument

We leverage the `_system_type` of `RelationshipManager` to determine
what `DofMap` from what system we should attach

Refs #16144

lindsayad committed:
Program DofMap for arbitrary number of Systems

Before I believe we were actually adding algebraic RelationshipManagers meant for
the nonlinear System/DofMap to the auxiliary System/DofMap and visa
versa. Now we should be adding individual relationship managers tailored
to each System/DofMap, e.g. we clone the "nonlinear" relationship
manager to make distinct relationship managers for each System/DofMap.
Of course as I'm writing this I'm realizing that we will now add the
same level of algebraic ghosting to all Systems/DofMaps (which was actually
already noted in #14536).

lindsayad commented:
@fdkong this doesn't address the geometric issue from #16144 but what do you think about https://github.com/idaholab/moose/pull/16619/commits/70f2f9d05e41c9938daaaa8f865064aa503121dc ? I think it is doing the algebraic version of what you already did for geometric, where you sure we were cloning new RMs for the displaced mesh. In https://github.com/idaholab/moose/pull/16619/commits/70f2f9d05e41c9938daaaa8f865064aa503121dc I've introduced cloning of algebraic relationship managers for the different `System/DofMaps`.

Doing this work has made me remember what I noted in #14536 which is that we have the same level of algebraic ghosting for all `Systems/DofMaps` so we may be ghosting dofs in `NumericVectors` that we don't need to.

fdkong commented:
An alternative solution is to have a PBC member variable pointer  in RM. And this PBC pointer is set from AddPeriodicBCAction.  All geometric RMs from nonlinear system and displaced system (if it exists) will then know PBC information. In such a way, we do not touch `dofmap` that is an algebraic  concept.

lindsayad commented:
> An alternative solution is to have a PBC member variable pointer in RM. And this PBC pointer is set from AddPeriodicBCAction. All geometric RMs from nonlinear system and displaced system (if it exists) will then know PBC information. In such a way, we do not touch `dofmap` that is an algebraic concept.

`GhostingFunctor` already has a `set_periodic_boundaries` member that I added that we could make use of. And I do make use of this in `ElementSideNeighborLayers::initFunctor`. The thing I don't understand is why we actually create `RelationshipManagers/GhostingFunctors` which just hold another `GhostingFunctor` functor. I think that makes our whole design a little harder and prone to error. When you init a `FunctorRelationshipManager` you have to make sure that you actually init the functor. It's really bizarre to me.

I still want to do the following two things:

1. Per discussion with @roystgnr add periodic boundaries to `MeshBase` when adding them to a `DofMap`. Then all periodic boundaries from all `DofMaps` would be added to the mesh held by `EquationSystems` and geometric ghosting functors could easily leverage that information through their `_mesh` member.
2. Clone algebraic ghosting functors/relationship managers in `SubProblem::addAlgebraicGhostingFunctor` and initialize those algebraic ghosting functors with the correct `DofMaps`.

lindsayad committed:
Add late algebraic/coupling before init to avoid need for reinit

lindsayad committed:
Revert "We can not rely on dofmap for periodic BCs because dof_map is set by only coupling RM"

This reverts commit d84bf9569e590aa9caac76542670ade9a553eb57.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/657214/) on 7ddb0cd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16619/site)

This comment will be updated on new commits.

lindsayad committed:
Always add algebraic relationship managers

Previously I had some crazy logic where I wouldn't add an algebraic RM
if it was also a coupling RM. This resulted in errors on a heat
conduction problem which required ghosting in the aux system...but since
I no longer individually add aux specific RMs where I remove the
coupling designation, I was no longer getting ghosting in the aux
system. Now that we clone RMs and apply the correct DofMaps, I no longer
have to worry about applying the wrong dof coupling and I can safely
always algebraic RMs

lindsayad committed:
The RelationshipManager added to the mesh shall have a DofMap

More specifically a RelationshipManager that does geometric ghosting
will have a DofMap applied to it at the time we add late geometric
ghosting functors, algebraic ghosting functors, and coupling functors.
In the case of periodic BCs on a distributed mesh, this will only help
you if you mark the RM responsible for your highest level of geometric
ghosting as a late addition. Otherwise your elements will be deleted at
the end of mesh setup

lindsayad committed:
Use _attach_geometric_early in determining >= comparisons

A Relationship Manager that must be attached late will be regarded as >=
than one that is attached early, because the former requires that *all*
elements be geometrically ghosted through the mesh setup phase.

lindsayad committed:
Track undisp to disp RM clones

If we want to remove a ghosting functor for example, we must make sure
to track the displaced clone in order to actually remove the correct RM
from the displaced mesh/problem. Previously we were trying remove the
undisplaced RM from the displaced mesh/problem which would essentially
have been a no-op due to failed std::set::find for that object

loganharbour reviewed (dismissed):
👍🏼 looks good to me and definitely fixes my issues. Thanks for getting onto this so quickly, I appreciate it.

Hopefully we should get bison passing today and can get this in this evening.

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -2083,112 +2089,101 @@ MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type,
+    else // rm_type is algebraic or coupling
+    {
+      // Now we've built the problem, so we can use it
+      auto & problem = _executioner->feProblem();
```
Should we check if _executioner exists? If it does not exist, we could error out

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -2083,112 +2089,101 @@ MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type,
-            problem.addAlgebraicGhostingFunctor(*rm, /*to_mesh = */ false);
-        }
+        else if (rm_type == Moose::RelationshipManagerType::ALGEBRAIC)
+          problem.addAlgebraicGhostingFunctor(*rm, /*to_mesh = */ false);
```
It is a great refactor of `MooseApp::attachRelationshipManagers`. The logic is so simple and also very easy to understand

fdkong commented on framework/src/problems/SubProblem.C:
```diff
@@ -922,9 +923,22 @@ SubProblem::addAlgebraicGhostingFunctor(GhostingFunctor & algebraic_gf, bool to_
 {
   EquationSystems & eq = es();
   auto n_sys = eq.n_systems();
+  eq.get_system(0).get_dof_map().add_algebraic_ghosting_functor(algebraic_gf, to_mesh);
```
Should we check if `n_sys>0`?

fdkong commented on framework/src/problems/SubProblem.C:
```diff
@@ -933,18 +947,57 @@ SubProblem::addAlgebraicGhostingFunctor(std::shared_ptr<GhostingFunctor> algebra
 
-  for (MooseIndex(n_sys) i = 0; i < n_sys; ++i)
-    eq.get_system(i).get_dof_map().add_algebraic_ghosting_functor(algebraic_gf, to_mesh);
+  eq.get_system(0).get_dof_map().add_algebraic_ghosting_functor(algebraic_gf, to_mesh);
```
The code in `SubProblem::addAlgebraicGhostingFunctor(shared pointer, bool ) `is very similar to `SubProblem::addAlgebraicGhostingFunctor(&, bool)`. Could we make a little helper, and reuse the code?

fdkong commented on framework/src/relationshipmanagers/RelationshipManager.C:
```diff
@@ -122,7 +110,13 @@ RelationshipManager::isCoupling(Moose::RelationshipManagerType input_rm)
+  // think of this as being >= because we if we are attaching late, then we are asking that *all*
+  // elements be geometrically ghosted during the initial mesh preparation phase. We will only allow
+  // remote elements to be deleted later on after addition of late geomeric ghosting functors (at
+  // the same time as algebraic and coupling)
```
Right, it is what I want because my mesh is generated in parallel and I am not in hurry to delete extra elements

fdkong commented on test/tests/dgkernels/2d_diffusion_dg/no_mallocs_with_action.i:
```diff
@@ -85,6 +85,10 @@
+  [console]
+    type = Console
+    system_info = 'framework mesh aux nonlinear relationship execution'
+  []
```
I have a dream that these  infos are printed out by default

fdkong reviewed (changes requested):
Looks nice! Minor comments

fdkong commented on framework/include/base/MooseApp.h:
```diff
@@ -708,11 +708,13 @@ class MooseApp : public ConsoleStreamInterface, public libMesh::ParallelObject
   /**
    * Purge this relationship manager from meshes and DofMaps and finally from us
    */
   void removeRelationshipManager(std::shared_ptr<RelationshipManager> relationship_manager);
```
It would be a big plus if you could add a comments about what siuation we need to remove RMs.

In my head, I alway think adding and adding. I do not have use cases where I need to remove RMs.

lindsayad committed:
Address review

lindsayad commented:
Patching THM here: https://gitlab.software.inl.gov/idaholab/thm/-/merge_requests/449

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/658389/) on 880bab5 : invalidated by @lindsayad

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/658389/) on 880bab5 : invalidated by @lindsayad

fdkong commented:
This made RMs so much easier to understand for me
